lib/spack/spack/traverse.py

```diff
@@ -205,16 +275,39 @@ def traverse_breadth_first_with_visitor(specs, visitor):
         visitor: object that implements accept and neighbors interface, see
             for example BaseVisitor.
     """
-    queue = root_specs(specs)
+    queue = with_artificial_edges(specs)
     while len(queue) > 0:
-        node = queue.pop(0)
+        edge = queue.pop(0)
 
         # If the visitor doesn't accept the node, we don't traverse it further.
-        if not visitor.accept(node):
+        if not visitor.accept(edge):
+            continue
+
+        for e in visitor.neighbors(edge):
+            queue.append(EdgeAndDepth(e, edge.depth + 1))
+
+
+def traverse_depth_first_with_visitor(edges, visitor):
+    """Traverse a DAG in depth-first fashion using a visitor, starting from
+    a list of edges. Note that typically DFS would take a vertex not a list of edges,
+    but the API is like this so we don't have to create an artificial root node when
+    traversing from multiple roots in a DAG.
+
+    Arguments:
+        edges (list): List of EdgeAndDepth instances
+        visitor: class instance implementing accept(), pre() and neigbors()
```

Comment:
I get your point, but I disagree, because there are instances where the `traverse_nodes` interface is too limiting, for example because `deptype` is not constant but depends on depth or the particular node. Also, the composability aspect is nice, and in principle it's possible to unit test the behavior of a (more complex) visitor.